algorithm/evocoin.c

```diff
@@ -0,0 +1,346 @@
+	sph_luffa512_context    luffa1;
+	sph_cubehash512_context cubehash1;
+	sph_shavite512_context  shavite1;
+	sph_simd512_context     simd1;
+	sph_echo512_context     echo1;
+} Xhash_context_holder;
+
+static Xhash_context_holder base_contexts;
+
+
+static void init_Xhash_contexts()
+{
+	sph_blake512_init(&base_contexts.blake1);
+	sph_bmw512_init(&base_contexts.bmw1);
+	sph_groestl512_init(&base_contexts.groestl1);
+	sph_skein512_init(&base_contexts.skein1);
+	sph_jh512_init(&base_contexts.jh1);
+	sph_keccak512_init(&base_contexts.keccak1);
+	sph_luffa512_init(&base_contexts.luffa1);
+	sph_cubehash512_init(&base_contexts.cubehash1);
+	sph_shavite512_init(&base_contexts.shavite1);
+	sph_simd512_init(&base_contexts.simd1);
+	sph_echo512_init(&base_contexts.echo1);
+}
+
+
+uint32_t getCurrentAlgoSeq(uint32_t current_time, uint32_t base_time) {
+	return (current_time - base_time) / (60 * 60 * 24);
+}
+
+void swap(uint8_t *a, uint8_t *b) {
+	uint8_t __tmp = *a;
+	*a = *b;
+	*b = __tmp;
+}
+
+void initPerm(uint8_t n[], uint8_t count) {
+	int i;
+	for (i = 0; i<count; i++)
+		n[i] = i;
+}
+
+int nextPerm(uint8_t n[], uint32_t count) {
+	uint32_t tail, i, j;
+
+	if (count <= 1)
+		return 0;
+
+	for (i = count - 1; i>0 && n[i - 1] >= n[i]; i--);
+	tail = i;
+
+	if (tail > 0) {
+		for (j = count - 1; j>tail && n[j] <= n[tail - 1]; j--);
+		swap(&n[tail - 1], &n[j]);
+	}
+
+	for (i = tail, j = count - 1; i<j; i++, j--)
+		swap(&n[i], &n[j]);
+
+	return (tail != 0);
+}
+
+
+void getAlgoString(char *str, uint32_t count)
+{
+	uint8_t algoList[HASH_FUNC_COUNT];
+	char s[100];
+	char *sptr;
+
+	initPerm(algoList, HASH_FUNC_COUNT);
+
+	int j;
+
+	int k;
+	for (k = 0; k < count; k++) {
+		nextPerm(algoList, HASH_FUNC_COUNT);
+	}
+
+	sptr = str;
+	for (j = 0; j < HASH_FUNC_COUNT; j++) {
+		if (algoList[j] >= 10)
+			sprintf(sptr, "%c", 'A' + (algoList[j] - 10));
+		else
+			sprintf(sptr, "%u", algoList[j]);
+		sptr++;
+	}
+	*sptr = 0;
+}
+
+
+void evocoin_twisted_code(char *result, const char *ntime, char *code)
+{
+	unsigned char bin[4];
+	uint32_t h32, *be32 = (uint32_t *)bin;
+	hex2bin(bin, ntime, 4);
+	h32 = be32toh(*be32);
+	uint32_t count = getCurrentAlgoSeq(h32, INITIAL_DATE);
+	getAlgoString(code, count);
+	sprintf(result, "_%d_%s_", count, code);
+}
+
+
+static inline void xhash(void *state, const void *input , const char* ntime)
+{
+	init_Xhash_contexts();
+
+	Xhash_context_holder ctx;
+
+	uint32_t hashA[16], hashB[16];
+	memcpy(&ctx, &base_contexts, sizeof(base_contexts));
+	
+	char completeCode[64];
+	char resultCode[HASH_FUNC_COUNT + 1];
+	evocoin_twisted_code(completeCode, ntime, resultCode);
+
+	int i;
+	const void *in;
+	void *out;
+
+
+	for (i = 0; i < strlen(resultCode); i++) {
+		char elem = resultCode[i];
+		uint8_t idx;
+		if (elem >= 'A')
+			idx = elem - 'A' + 10;
+		else
+			idx = elem - '0';
+
+		int size;
+
+		if (i == 0) {
+			in = input;
+			size = 80;
+			out = hashA;
+		}
+		else {
+			if (out == hashA) {
```

Comment:
no need this hashA/hashB swaps, just use one hash[16] variable: 

    if i == 0
        in=input; size=80;
    else 
        in = hash; size = 64;